File: python/src/tdcv2/date/gen.py
"""``<gen type="date">`` — a moment, a range of them, or a birthday.

One generator covers four shapes because they are four ways of asking the same question: a fixed
date, a range written as ``from``/``to``, ``today``/``now`` read from the clock, and ``birth``,
which is a range expressed in ages rather than dates. Splitting them apart would make the common
case — "some date in the nineties" — a choice between generators.

Precision decides what is drawn, not just how it prints. At day precision a whole day is one
outcome, so a range of ten years has 3653 of them; at millisecond precision the same range has
315 billion, and every row lands on a different instant. That is the difference between an order
date and an event timestamp, and only the config knows which one is wanted.
"""

from __future__ import annotations

import math
from collections.abc import Callable
from dataclasses import dataclass

from ..prng.prng import Sfc32
from . import calendar, parse
from .formatter import format_date_time
from .plain import (
    MS_PER_SECOND,
    DateError,
    PlainDateTime,
    Precision,
    from_epoch_day,
    from_epoch_millis,
    subtract_years,
    to_epoch_day,
    to_epoch_millis,
)

DEFAULT_START = "1970-01-01"
DEFAULT_FORMAT = "L"

_DEFAULT_OLDEST = 80
_DEFAULT_YOUNGEST = 10


@dataclass(frozen=True, slots=True)
class Plan:
    """What a run of this generator will draw from, decided once."""

    fixed: PlainDateTime | None
    start: PlainDateTime | None
    end: PlainDateTime | None
    precision: Precision
    format: str
    locale: str


def generate(
    attrs: dict[str, str],
    count: int,
    locale: str,
    now: int,
    prng: Sfc32,
    instants_out: list[int | None] | None = None,
) -> list[str]:
    """``count`` formatted dates.

    ``instants_out``, when given, receives the epoch millis behind each rendered cell — the value
    the generator actually produced, before ``format=`` turned it into one locale's spelling of
    it. A column another one measures from asks for this; everything else passes nothing and the
    list is never built.
    """
    plan = build_plan(attrs, locale, now)
    out: list[str] = []
    for _ in range(count):
        value = plan.fixed if plan.fixed is not None else _pick(plan, prng)
        if instants_out is not None:
            instants_out.append(to_epoch_millis(value))
        out.append(format_date_time(value, plan.format, plan.locale))
    return out


def build_plan(attrs: dict[str, str], locale: str, now: int) -> Plan:
    fmt = attrs.get("format", DEFAULT_FORMAT)
    loc = attrs.get("local", locale)
    value = attrs["value"].strip() if attrs.get("value") is not None else None

    if value == "today":
        return Plan(
            from_epoch_millis(now).start_of_day(),
            None,
            None,
            parse_precision(attrs.get("precision"), Precision.DAY),
            fmt,
            loc,
        )

    if value == "now":
        return Plan(
            from_epoch_millis(now),
            None,
            None,
            parse_precision(attrs.get("precision"), Precision.MILLISECOND),
            fmt,
            loc,
        )

    if value == "birth":
        # Ages, not dates: "between 10 and 80 years old" stays true as the clock moves, where a
        # pair of fixed dates would quietly age the whole population by a year every January.
        oldest = _age(attrs.get("oldest"), _DEFAULT_OLDEST, "oldest")
        youngest = _age(attrs.get("youngest"), _DEFAULT_YOUNGEST, "youngest")
        if youngest > oldest:
            raise DateError("date generator: youngest must be less than or equal to oldest")
        start = from_epoch_millis(subtract_years(now, oldest))
        end = from_epoch_millis(subtract_years(now, youngest))
        return _range_plan(
            parse.Range(parse.Parsed(start, False), parse.Parsed(end, False)),
            attrs,
            fmt,
            loc,
            Precision.DAY,
        )

    if attrs.get("from") is not None or attrs.get("to") is not None:
        # `from=` alone is an OPEN axis — legal when the range is WALKED, and the plan carries
        # only a start. `date_axis` reads `end` as None and never wraps; a DRAWN date with one
        # end is still refused, by TDC150.
        if attrs.get("from") is not None and attrs.get("to") is None:
            start = parse.date_time(attrs["from"])
            return Plan(
                None,
                start.value,
                None,
                parse_precision(
                    attrs.get("precision"),
                    Precision.MILLISECOND if start.has_time else Precision.DAY,
                ),
                fmt,
                loc,
            )
        if attrs.get("from") is None or attrs.get("to") is None:
            raise DateError('date generator: "from" and "to" must be provided together')
        bounds = parse.Range(parse.date_time(attrs["from"]), parse.date_time(attrs["to"]))
        return _range_plan(bounds, attrs, fmt, loc)

    if attrs.get("range") is not None:
        return _range_plan(parse.value_range(attrs["range"]), attrs, fmt, loc)

    if value:
        if ".." in value:
            return _range_plan(parse.value_range(value), attrs, fmt, loc)
        parsed = parse.date_time(value)
        return Plan(
            parsed.value,
            None,
            None,
            parse_precision(
                attrs.get("precision"),
                Precision.MILLISECOND if parsed.has_time else Precision.DAY,
            ),
            fmt,
            loc,
        )

    # Nothing said at all: the whole span from the epoch to right now, by day.
    bounds = parse.Range(parse.date_time(DEFAULT_START), parse.Parsed(from_epoch_millis(now), True))
    return _range_plan(bounds, attrs, fmt, loc, Precision.DAY)


_MS_PER_WEEK = 7 * 86_400_000


def _is_open_axis(attrs: dict[str, str]) -> bool:
    """True when a range was written with only its START — an axis with no end."""
    return (
        attrs.get("from") is not None
        and attrs.get("to") is None
        and attrs.get("range") is None
        and not (attrs.get("value") or "")
    )


@dataclass(frozen=True, slots=True)
class Axis:
    """A walkable date range: how many steps it holds, and what the k-th is.

    ``size`` is ``None`` for an OPEN axis — ``from=`` with no end — which never wraps, because
    there is nothing to wrap at.
    """

    size: int | None
    at: Callable[[int], str]


def date_axis(attrs: dict[str, str], locale: str, now: int) -> Axis:
    """A date range as a walkable axis.

    The range is never expanded into a list. A century stepped by the second is three billion
    values and the streaming engine promises bounded memory whatever the config says — so each
    date is ``start + k × step``, measured from the START rather than accumulated, which is what
    keeps a clamped February from dragging every later month back with it.
    """
    parsed = calendar.parse_step(attrs.get("step"))
    step = parsed.step if parsed.step is not None else calendar.DEFAULT_STEP
    keep = calendar.parse_weekdays(attrs.get("weekdays"))
    plan = build_plan(attrs, locale, now)

    def render(value: PlainDateTime) -> str:
        return format_date_time(value, plan.format, plan.locale)

    if plan.start is None:
        fixed = plan.fixed
        if fixed is None:
            raise DateError("date generator: invalid generation plan")
        return Axis(1, lambda _k: render(fixed))
    start = plan.start

    # `weekdays=` keeps only some of the candidates, so the k-th KEPT one is wanted rather than
    # the k-th candidate. Which candidates match repeats on a cycle — one week's worth of steps —
    # so the offsets are found once and then indexed, instead of scanning from the beginning for
    # every row.
    offsets: list[int] = []
    per_cycle = 0
    if keep is not None:
        per_cycle = _MS_PER_WEEK // math.gcd(step.ms, _MS_PER_WEEK) if step.ms > 0 else 7
        offsets = [
            i
            for i in range(per_cycle)
            if calendar.weekday_of(calendar.add_step(start, step, i)) in keep
        ]

    def candidate_at(k: int) -> PlainDateTime:
        if keep is None or not offsets:
            return calendar.add_step(start, step, k)
        cycles, within = divmod(k, len(offsets))
        return calendar.add_step(start, step, cycles * per_cycle + offsets[within])

    if _is_open_axis(attrs) or plan.end is None:
        return Axis(None, lambda k: render(candidate_at(k)))

    candidates = calendar.steps_between(start, plan.end, step)
    if keep is not None and offsets:
        size = max(
            1,
            candidates // per_cycle * len(offsets)
            + sum(1 for o in offsets if o < candidates % per_cycle),
        )
    elif keep is not None:
        size = 1
    else:
        size = candidates
    return Axis(size, lambda k: render(candidate_at(k)))


def parse_precision(raw: str | None, fallback: Precision = Precision.DAY) -> Precision:
    if raw is None:
        return fallback
    for precision in Precision:
        if raw == precision.value:
            return precision
    raise DateError(
        f'date generator: unsupported precision "{raw}" (supported: day, second, millisecond)'
    )


def render_birthday(attrs: dict[str, str], locale: str, now: int, prng: Sfc32) -> str:
    """The ``person.b_day`` template, which is this generator under another name."""
    forwarded = {
        "value": "birth",
        "precision": attrs.get("precision", "millisecond"),
        **{k: attrs[k] for k in ("oldest", "youngest", "format", "local") if k in attrs},
    }
    return generate(forwarded, 1, locale, now, prng)[0]


def render_date_range(attrs: dict[str, str], locale: str, now: int, prng: Sfc32) -> str:
    """The ``date.range`` template, which takes the older dotted spelling of a range."""
    raw = attrs.get("range", "")
    bounds = parse.legacy_range(raw)
    forwarded = {
        "from": _serialize(bounds.start.value),
        "to": _serialize(bounds.end.value),
        "precision": attrs.get("precision", "day"),
        **{k: attrs[k] for k in ("format", "local") if k in attrs},
    }
    return generate(forwarded, 1, locale, now, prng)[0]


def _range_plan(
    bounds: parse.Range,
    attrs: dict[str, str],
    fmt: str,
    locale: str,
    fallback: Precision | None = None,
) -> Plan:
    # A range written with a time in it means the config cares about the time, so the default
    # precision follows what was written rather than a fixed choice.
    has_time = bounds.start.has_time or bounds.end.has_time
    default = fallback or (Precision.MILLISECOND if has_time else Precision.DAY)
    return Plan(
        None,
        bounds.start.value,
        bounds.end.value,
        parse_precision(attrs.get("precision"), default),
        fmt,
        locale,
    )


def _pick(plan: Plan, prng: Sfc32) -> PlainDateTime:
    if plan.start is None or plan.end is None:
        raise DateError("date generator: range plan is incomplete")

    if plan.precision is Precision.DAY:
        a = to_epoch_day(plan.start)
        b = to_epoch_day(plan.end)
        return from_epoch_day(_inclusive(prng, min(a, b), max(a, b)))

    divisor = MS_PER_SECOND if plan.precision is Precision.SECOND else 1
    a = to_epoch_millis(plan.start) // divisor
    b = to_epoch_millis(plan.end) // divisor
    return from_epoch_millis(_inclusive(prng, min(a, b), max(a, b)) * divisor)


def _inclusive(prng: Sfc32, minimum: int, maximum: int) -> int:
    return math.floor(prng.next() * (maximum - minimum + 1) + minimum)


def _age(raw: str | None, fallback: int, name: str) -> int:
    if raw is None:
        return fallback
    try:
        value = int(raw.strip())
    except ValueError:
        raise DateError(f"date generator: {name} must be an integer from 0 to 150") from None
    if value < 0 or value > 150:
        raise DateError(f"date generator: {name} must be an integer from 0 to 150")
    return value


def _serialize(value: PlainDateTime) -> str:
    return (
        f"{value.year:04d}-{value.month:02d}-{value.day:02d}"
        f"T{value.hour:02d}:{value.minute:02d}:{value.second:02d}.{value.millisecond:03d}"
    )

File: rust/src/date/gen.rs
//! `<gen type="date" .../>` and the `person.b_day` template behind it.
//!
//! A plan is built once from the attributes, then each value is one draw against
//! it. Two kinds: a fixed instant (`today`, `now`, a single date) that takes no
//! draw at all, and a range that takes exactly one.
//!
//! Precision decides what the draw is over — days, seconds or milliseconds — and
//! it is not cosmetic. A range drawn by day and the same range drawn by
//! millisecond both look like dates once formatted, and they disagree.

use std::collections::BTreeMap;

use super::{
    calendar, floor_div, format, from_epoch_day, from_epoch_millis, parse, subtract_utc_years,
    to_epoch_day, to_epoch_millis, PlainDateTime, MS_PER_DAY,
};
use crate::engine::{invalid, EngineResult};
use crate::prng::Sfc32;

const DEFAULT_START: &str = "1970-01-01";
const DEFAULT_FORMAT: &str = "L";
const MS_PER_SECOND: i64 = 1000;

/// How finely the range is divided before a value is drawn from it.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum Precision {
    Day,
    Second,
    Millisecond,
}

struct Plan {
    fixed: Option<PlainDateTime>,
    start: PlainDateTime,
    end: PlainDateTime,
    /// True when the range was written with only its START. `end` is then a copy
    /// of `start` and means nothing; `date_axis` reads this and never wraps.
    open_end: bool,
    grain: Precision,
    format: String,
    locale: Option<String>,
}

pub fn generate(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
    count: usize,
    prng: &mut Sfc32,
) -> EngineResult<Vec<String>> {
    generate_into(attrs, locale, now_millis, count, prng, None)
}

/// `count` formatted dates, optionally keeping the value behind each one.
///
/// `instants`, when given, receives the epoch millis the generator actually
/// produced, before `format=` turned it into one locale's spelling of it. A
/// column another one measures from asks for this; everything else passes `None`
/// and nothing is collected.
pub fn generate_into(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
    count: usize,
    prng: &mut Sfc32,
    mut instants: Option<&mut Vec<Option<i64>>>,
) -> EngineResult<Vec<String>> {
    let plan = build_plan(attrs, locale, now_millis)?;
    format::check_format(&plan.format)?;
    let mut result = Vec::with_capacity(count);
    for _ in 0..count {
        let value = match plan.fixed {
            Some(fixed) => fixed,
            None => pick(&plan, prng),
        };
        if let Some(sink) = instants.as_deref_mut() {
            sink.push(Some(to_epoch_millis(value)));
        }
        result.push(format::format(
            value,
            Some(&plan.format),
            plan.locale.as_deref(),
        ));
    }
    Ok(result)
}

/// One value for `person.b_day`, which is a date generator wearing a template's
/// name.
pub fn birth_day(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
    prng: &mut Sfc32,
) -> EngineResult<String> {
    Ok(generate(&birth_attrs(attrs), locale, now_millis, 1, prng)?.remove(0))
}

/// `date.range` — a date generator addressed as a pack path, taking the older
/// `range="1990.01.01 - 2000.12.31"` spelling.
///
/// It is the `date` generator underneath, so the bounds are rewritten into the
/// form that generator reads rather than a second implementation being kept in
/// step with the first.
pub fn legacy_range(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
    count: usize,
    prng: &mut Sfc32,
) -> EngineResult<Vec<String>> {
    let raw = attrs.get("range").map(String::as_str).unwrap_or("");
    let Ok(range) = parse::legacy_range(raw) else {
        return invalid(&format!("date.range: invalid range attribute \"{raw}\""));
    };

    let mut rewritten = BTreeMap::new();
    rewritten.insert("from".to_string(), serialize(range.start.value));
    rewritten.insert("to".to_string(), serialize(range.end.value));
    rewritten.insert(
        "precision".to_string(),
        attrs
            .get("precision")
            .cloned()
            .unwrap_or_else(|| "day".to_string()),
    );
    copy(attrs, &mut rewritten, "format");
    copy(attrs, &mut rewritten, "local");
    generate(&rewritten, locale, now_millis, count, prng)
}

fn serialize(v: PlainDateTime) -> String {
    format!(
        "{:04}-{:02}-{:02}T{:02}:{:02}:{:02}.{:03}",
        v.year, v.month, v.day, v.hour, v.minute, v.second, v.millisecond
    )
}

/// `person.b_day` reaches the date generator with `value="birth"` and an
/// explicit millisecond precision.
///
/// The precision looks redundant next to a birth range measured in years, and it
/// is not: it is what the reference passes, so it is what decides the day.
fn birth_attrs(attrs: &BTreeMap<String, String>) -> BTreeMap<String, String> {
    let mut result = BTreeMap::new();
    result.insert("value".to_string(), "birth".to_string());
    result.insert(
        "precision".to_string(),
        attrs
            .get("precision")
            .cloned()
            .unwrap_or_else(|| "millisecond".to_string()),
    );
    for key in ["oldest", "youngest", "format", "local"] {
        copy(attrs, &mut result, key);
    }
    result
}

fn copy(from: &BTreeMap<String, String>, to: &mut BTreeMap<String, String>, key: &str) {
    if let Some(value) = from.get(key) {
        to.insert(key.to_string(), value.clone());
    }
}

fn build_plan(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
) -> EngineResult<Plan> {
    let format = attrs
        .get("format")
        .cloned()
        .unwrap_or_else(|| DEFAULT_FORMAT.to_string());
    let loc = match attrs.get("local") {
        Some(own) => Some(own.clone()),
        None => locale.map(str::to_string),
    };
    let value = attrs.get("value").map(|v| v.trim());
    let precision = attrs.get("precision").map(String::as_str);

    if value == Some("today") {
        return Ok(fixed(
            from_epoch_millis(now_millis).start_of_day(),
            parse_precision(precision, Precision::Day)?,
            format,
            loc,
        ));
    }

    if value == Some("now") {
        return Ok(fixed(
            from_epoch_millis(now_millis),
            parse_precision(precision, Precision::Millisecond)?,
            format,
            loc,
        ));
    }

    if value == Some("birth") {
        let oldest = age(attrs.get("oldest"), 80, "oldest")?;
        let youngest = age(attrs.get("youngest"), 10, "youngest")?;
        if youngest > oldest {
            return invalid("date generator: youngest must be less than or equal to oldest");
        }
        // The default is by DAY even though `person.b_day` asks for
        // milliseconds: the fallback belongs to `value="birth"` written out in a
        // config, and the template overrides it.
        return range_plan(
            from_epoch_millis(subtract_utc_years(now_millis, oldest)),
            from_epoch_millis(subtract_utc_years(now_millis, youngest)),
            precision,
            Precision::Day,
            format,
            loc,
        );
    }

    let from = attrs.get("from");
    let to = attrs.get("to");
    if from.is_some() || to.is_some() {
        // `from=` alone is an OPEN axis — legal when the range is WALKED, and the
        // plan carries only a start. `date_axis` reads `open_end` and never wraps;
        // a DRAWN date with one end is still refused, by TDC150.
        if let (Some(from), None) = (from, to) {
            let start = parse::date_time(from)?;
            return Ok(Plan {
                fixed: None,
                start: start.value,
                end: start.value,
                open_end: true,
                grain: parse_precision(
                    precision,
                    if start.has_time {
                        Precision::Millisecond
                    } else {
                        Precision::Day
                    },
                )?,
                format,
                locale: loc,
            });
        }
        let (Some(from), Some(to)) = (from, to) else {
            return invalid("date generator: \"from\" and \"to\" must be provided together");
        };
        return range_of(
            parse::date_time(from)?,
            parse::date_time(to)?,
            precision,
            format,
            loc,
        );
    }

    if let Some(raw) = attrs.get("range") {
        let parsed = parse::range(raw)?;
        return range_of(parsed.start, parsed.end, precision, format, loc);
    }

    if let Some(value) = value.filter(|v| !v.is_empty()) {
        if value.contains("..") {
            let parsed = parse::range(value)?;
            return range_of(parsed.start, parsed.end, precision, format, loc);
        }
        let one = parse::date_time(value)?;
        return Ok(fixed(
            one.value,
            parse_precision(
                precision,
                if one.has_time {
                    Precision::Millisecond
                } else {
                    Precision::Day
                },
            )?,
            format,
            loc,
        ));
    }

    // Nothing specified at all: the epoch up to right now. The upper bound carries
    // a time, but the fallback precision is still whole days — an unbounded
    // generator answers with a date, not a timestamp at 03:47. Routing this
    // through `range_of` let `has_time` pick Millisecond, and a millisecond draw
    // lands a day away from the reference's day draw often enough to fail.
    range_plan(
        parse::date_time(DEFAULT_START)?.value,
        from_epoch_millis(now_millis),
        precision,
        Precision::Day,
        format,
        loc,
    )
}

fn fixed(value: PlainDateTime, grain: Precision, format: String, locale: Option<String>) -> Plan {
    Plan {
        fixed: Some(value),
        start: PlainDateTime::default(),
        end: PlainDateTime::default(),
        open_end: false,
        grain,
        format,
        locale,
    }
}

fn range_of(
    start: parse::Parsed,
    end: parse::Parsed,
    precision: Option<&str>,
    format: String,
    locale: Option<String>,
) -> EngineResult<Plan> {
    // When neither bound carried a time, the range is over whole days — which is
    // why `range="2026-01-01..2026-01-31"` yields dates and not timestamps at
    // 03:47.
    let fallback = if start.has_time || end.has_time {
        Precision::Millisecond
    } else {
        Precision::Day
    };
    range_plan(start.value, end.value, precision, fallback, format, locale)
}

fn range_plan(
    start: PlainDateTime,
    end: PlainDateTime,
    precision: Option<&str>,
    fallback: Precision,
    format: String,
    locale: Option<String>,
) -> EngineResult<Plan> {
    Ok(Plan {
        fixed: None,
        start,
        end,
        open_end: false,
        grain: parse_precision(precision, fallback)?,
        format,
        locale,
    })
}

/// A date range as a walkable axis: how many steps it holds, and what the k-th is.
///
/// `size` is `None` for an OPEN axis — `from=` with no end. Requiring an end
/// meant working out what date the millionth day falls on in order to write it
/// down, when the end is simply `start + count x step`. Such an axis never wraps,
/// because there is nothing to wrap at.
///
/// The range is never expanded into a list. A century stepped by the second is
/// three billion values and the streaming engine promises bounded memory whatever
/// the config says — so each date is `start + k x step`, measured from the START
/// rather than accumulated, which is what keeps a clamped February from dragging
/// every later month back with it.
#[derive(Debug)]
pub struct Axis {
    pub size: Option<i64>,
    start: PlainDateTime,
    step: calendar::StepSpec,
    /// Which offsets within one cycle a `weekdays=` filter keeps. Empty when
    /// there is no filter.
    offsets: Vec<i64>,
    per_cycle: i64,
    format: String,
    locale: Option<String>,
}

impl Axis {
    /// The k-th value of the axis, rendered.
    pub fn at(&self, k: i64) -> String {
        format::format(self.value_at(k), Some(&self.format), self.locale.as_deref())
    }

    /// The k-th value of the axis, BEFORE `format=` turns it into one locale's
    /// spelling of it — what a column measuring from this one has to read.
    pub fn value_at(&self, k: i64) -> PlainDateTime {
        if self.offsets.is_empty() {
            calendar::add_step(self.start, self.step, k)
        } else {
            let n = self.offsets.len() as i64;
            let cycles = k.div_euclid(n);
            let within = self.offsets[k.rem_euclid(n) as usize];
            calendar::add_step(self.start, self.step, cycles * self.per_cycle + within)
        }
    }
}

fn gcd(a: i64, b: i64) -> i64 {
    if b == 0 {
        a
    } else {
        gcd(b, a % b)
    }
}

const MS_PER_WEEK: i64 = 7 * MS_PER_DAY;

/// Build the walkable axis a `order="sequential"` date reads.
pub fn date_axis(
    attrs: &BTreeMap<String, String>,
    locale: Option<&str>,
    now_millis: i64,
) -> EngineResult<Axis> {
    let step = calendar::parse_step(attrs.get("step").map(String::as_str))
        .unwrap_or(calendar::DEFAULT_STEP);
    let keep = calendar::parse_weekdays(attrs.get("weekdays").map(String::as_str));
    let plan = build_plan(attrs, locale, now_millis)?;
    format::check_format(&plan.format)?;

    if let Some(fixed) = plan.fixed {
        return Ok(Axis {
            size: Some(1),
            start: fixed,
            step: calendar::DEFAULT_STEP,
            offsets: Vec::new(),
            per_cycle: 1,
            format: plan.format,
            locale: plan.locale,
        });
    }

    // `weekdays=` keeps only some of the candidates, so the k-th KEPT one is
    // wanted rather than the k-th candidate. Which candidates match repeats on a
    // cycle — one week's worth of steps — so the offsets are found once and then
    // indexed, instead of scanning from the beginning for every row.
    let mut offsets: Vec<i64> = Vec::new();
    let mut per_cycle = 1i64;
    if let Some(keep) = keep {
        per_cycle = if step.ms > 0 {
            MS_PER_WEEK / gcd(step.ms, MS_PER_WEEK)
        } else {
            7
        };
        for i in 0..per_cycle {
            if keep[calendar::weekday_of(calendar::add_step(plan.start, step, i))] {
                offsets.push(i);
            }
        }
    }

    let size = if plan.open_end {
        None
    } else {
        let candidates = calendar::steps_between(plan.start, plan.end, step);
        Some(if offsets.is_empty() {
            candidates
        } else {
            (candidates / per_cycle * offsets.len() as i64
                + offsets
                    .iter()
                    .filter(|o| **o < candidates % per_cycle)
                    .count() as i64)
                .max(1)
        })
    };

    Ok(Axis {
        size,
        start: plan.start,
        step,
        offsets,
        per_cycle,
        format: plan.format,
        locale: plan.locale,
    })
}

fn pick(plan: &Plan, prng: &mut Sfc32) -> PlainDateTime {
    if plan.grain == Precision::Day {
        let a = to_epoch_day(plan.start);
        let b = to_epoch_day(plan.end);
        return from_epoch_day(inclusive(prng, a.min(b), a.max(b)));
    }

    let divisor = if plan.grain == Precision::Second {
        MS_PER_SECOND
    } else {
        1
    };
    let lo = floor_div(to_epoch_millis(plan.start), divisor);
    let hi = floor_div(to_epoch_millis(plan.end), divisor);
    from_epoch_millis(inclusive(prng, lo.min(hi), lo.max(hi)) * divisor)
}

/// One draw, inclusive of both ends.
fn inclusive(prng: &mut Sfc32, min: i64, max: i64) -> i64 {
    (prng.next() * (max - min + 1) as f64 + min as f64).floor() as i64
}

pub fn parse_precision(raw: Option<&str>, fallback: Precision) -> EngineResult<Precision> {
    match raw {
        None => Ok(fallback),
        Some("day") => Ok(Precision::Day),
        Some("second") => Ok(Precision::Second),
        Some("millisecond") => Ok(Precision::Millisecond),
        Some(other) => invalid(&format!(
            "date generator: unsupported precision \"{other}\" \
             (supported: day, second, millisecond)"
        )),
    }
}

/// The birth ages, checked without generating anything.
///
/// Whole numbers in a plausible range, and the older bound actually older — a
/// config that has them the wrong way round asks for an empty span and gets no
/// dates at all.
pub fn check_birth_ages(attrs: &BTreeMap<String, String>) -> EngineResult<()> {
    let oldest = age(attrs.get("oldest"), 80, "oldest")?;
    let youngest = age(attrs.get("youngest"), 10, "youngest")?;
    if youngest > oldest {
        return invalid("date generator: youngest must be less than or equal to oldest");
    }
    Ok(())
}

fn age(raw: Option<&String>, fallback: i32, name: &str) -> EngineResult<i32> {
    let Some(raw) = raw else {
        return Ok(fallback);
    };
    match raw.trim().parse::<i32>() {
        Ok(value) if (0..=150).contains(&value) => Ok(value),
        _ => invalid(&format!(
            "date generator: {name} must be an integer from 0 to 150"
        )),
    }
}
